src/main/java/co/cask/gcp/bigquery/BigQuerySink.java

```diff
@@ -442,4 +451,104 @@ private void decodeSimpleTypes(JsonObject json, String name, StructuredRecord in
         break;
     }
   }
+
+  /**
+   * Validates output schema against bigquery table schema. It throws {@link IllegalArgumentException}
+   * if the output schema has more fields than bigquery table or output schema field types does not match bigquery
+   * column types.
+   */
+  private void validateSchema(BigQuerySinkConfig config) throws IOException {
+    BigQuery bigquery;
+    BigQueryOptions.Builder bigqueryBuilder = BigQueryOptions.newBuilder();
+    String serviceAccountFilePath = config.getServiceAccountFilePath();
+    if (serviceAccountFilePath != null) {
+      bigqueryBuilder.setCredentials(loadServiceAccountCredentials(serviceAccountFilePath));
+    }
+    String project = config.getProject();
+    bigqueryBuilder.setProjectId(project);
+    bigquery = bigqueryBuilder.build().getService();
+
+    TableId id = TableId.of(project, config.dataset, config.table);
+    Table table = bigquery.getTable(id);
+    if (table == null) {
+      // Table does not exist, so no further validation is required.
+      return;
+    }
+
+    com.google.cloud.bigquery.Schema bqSchema = table.getDefinition().getSchema();
+    if (bqSchema == null) {
+      // Table is created without schema, so no further validation is required.
+      return;
+    }
+
+    FieldList bqFields = bqSchema.getFields();
+    List<Schema.Field> outputSchemaFields = Schema.parseJson(config.schema).getFields();
+
+    if (bqFields.size() != outputSchemaFields.size()) {
+      throw new IllegalArgumentException(
+        "Output schema does not match BigQuery table schema. Output schema must be compatible with " +
```

Comment:
We should be able to give a more detailed message. If there is a required column in BQ that is not in the output schema, the message should look something like:

The output schema does not match the BigQuery table schema for '[dataset].[table]'. The table requires column 'xyz', which is not in the output schema.

If the output schema contains a field that is not in the BQ table:

The output schema does not match the BigQuery table schema for '[dataset].[table]'. The table does not contain the 'xyz' column.